File: Tema5/dockerlexer/java/Parser.java
package dockerlexer;

import java.util.*;
import dockerlexer.Lexer.Token;
import dockerlexer.Lexer.TokenType;

/**
 * Parser para Dockerfile. Consume los tokens del Lexer y construye un AST
 * como lista de nodos (Map), siguiendo la misma gramatica que las demas
 * implementaciones (Python, Rust).
 */
public class Parser {

    public static class ParseException extends RuntimeException {
        public ParseException(String msg) { super(msg); }
    }

    public static class Node {
        public String instruction;
        public List<String> args = new ArrayList<>();
        public String form; // puede ser null

        public String toJson() {
            StringBuilder sb = new StringBuilder();
            sb.append("{");
            sb.append("\"instruction\": \"").append(escape(instruction)).append("\", ");
            sb.append("\"args\": [");
            for (int i = 0; i < args.size(); i++) {
                if (i > 0) sb.append(", ");
                sb.append("\"").append(escape(args.get(i))).append("\"");
            }
            sb.append("]");
            if (form != null) {
                sb.append(", \"form\": \"").append(escape(form)).append("\"");
            }
            sb.append("}");
            return sb.toString();
        }

        static String escape(String s) {
            if (s == null) return "";
            return s.replace("\\", "\\\\").replace("\"", "\\\"");
        }
    }

    private final List<Token> tokens;
    private int pos = 0;

    public Parser(List<Token> tokens) {
        this.tokens = tokens;
    }

    private Token peek() {
        return tokens.get(pos);
    }

    private Token advance() {
        return tokens.get(pos++);
    }

    public List<Node> parse() {
        List<Node> ast = new ArrayList<>();
        boolean seenFrom = false;

        while (peek().type != TokenType.EOF) {
            Token tok = peek();

            if (tok.type == TokenType.COMMENT) {
                advance();
                expectOptionalNewline();
                continue;
            }

            if (tok.type == TokenType.UNKNOWN) {
                throw new ParseException(
                    "Instruccion desconocida '" + tok.value + "' en linea " + tok.line);
            }

            if (tok.type == TokenType.INSTRUCTION) {
                Node node = parseInstruction();
                if (node.instruction.equals("FROM")) {
                    seenFrom = true;
                } else if (!seenFrom && !node.instruction.equals("ARG")) {
                    throw new ParseException(
                        "La instruccion '" + node.instruction + "' aparece antes de FROM (linea "
                        + tok.line + ")");
                }
                ast.add(node);
                continue;
            }

            if (tok.type == TokenType.NEWLINE) {
                advance();
                continue;
            }

            throw new ParseException("Token inesperado " + tok.type + " en linea " + tok.line);
        }

        return ast;
    }

    private void expectOptionalNewline() {
        if (peek().type == TokenType.NEWLINE) advance();
    }

    private Node parseInstruction() {
        Token instrTok = advance();
        Node node = new Node();
        node.instruction = instrTok.value;

        if (peek().type == TokenType.LBRACKET) {
            parseExecForm(node);
        } else {
            parseShellOrKvForm(node);
        }

        expectOptionalNewline();
        return node;
    }

    private void parseExecForm(Node node) {
        advance(); // consume '['
        while (peek().type != TokenType.RBRACKET) {
            Token tok = peek();
            if (tok.type == TokenType.STRING) {
                node.args.add(tok.value);
                advance();
            } else if (tok.type == TokenType.COMMA) {
                advance();
            } else if (tok.type == TokenType.NEWLINE) {
                advance();
            } else {
                throw new ParseException(
                    "Token inesperado " + tok.type + " dentro de forma exec, linea " + tok.line);
            }
        }
        advance(); // consume ']'
        node.form = "exec";
    }

    private void parseShellOrKvForm(Node node) {
        boolean isKv = false;

        while (peek().type == TokenType.IDENTIFIER
                || peek().type == TokenType.EQUALS
                || peek().type == TokenType.STRING) {
            Token tok = advance();
            if (tok.type == TokenType.EQUALS) {
                isKv = true;
                continue;
            }
            node.args.add(tok.value);
        }

        node.form = isKv ? "key_value" : "shell";
    }

    public static List<Node> parseDockerfile(String text) {
        List<Token> tokens = Lexer.tokenize(text);
        Parser parser = new Parser(tokens);
        return parser.parse();
    }
}

File: Tema5/dockerlexer/rust/src/parser.rs
// Parser para Dockerfile. Consume los tokens de lexer.rs y construye un
// AST como lista de nodos, siguiendo la misma gramatica que las demas
// implementaciones (Python, Java).

use crate::lexer::{tokenize, Token, TokenType};

#[derive(Debug)]
pub struct Node {
    pub instruction: String,
    pub args: Vec<String>,
    pub form: Option<String>,
}

impl Node {
    pub fn to_json(&self) -> String {
        let args_json: Vec<String> = self
            .args
            .iter()
            .map(|a| format!("\"{}\"", escape(a)))
            .collect();

        let mut s = format!(
            "{{\"instruction\": \"{}\", \"args\": [{}]",
            escape(&self.instruction),
            args_json.join(", ")
        );

        if let Some(form) = &self.form {
            s.push_str(&format!(", \"form\": \"{}\"", escape(form)));
        }

        s.push('}');
        s
    }
}

fn escape(s: &str) -> String {
    s.replace('\\', "\\\\").replace('"', "\\\"")
}

#[derive(Debug)]
pub struct ParseError(pub String);

pub struct Parser {
    tokens: Vec<Token>,
    pos: usize,
}

impl Parser {
    pub fn new(tokens: Vec<Token>) -> Self {
        Parser { tokens, pos: 0 }
    }

    fn peek(&self) -> &Token {
        &self.tokens[self.pos]
    }

    fn advance(&mut self) -> Token {
        let tok = self.tokens[self.pos].clone();
        self.pos += 1;
        tok
    }

    pub fn parse(&mut self) -> Result<Vec<Node>, ParseError> {
        let mut ast = Vec::new();
        let mut seen_from = false;

        while self.peek().ttype != TokenType::Eof {
            let tok = self.peek().clone();

            match tok.ttype {
                TokenType::Comment => {
                    self.advance();
                    self.expect_optional_newline();
                    continue;
                }
                TokenType::Unknown => {
                    return Err(ParseError(format!(
                        "Instruccion desconocida '{}' en linea {}",
                        tok.value, tok.line
                    )));
                }
                TokenType::Instruction => {
                    let node = self.parse_instruction()?;
                    if node.instruction == "FROM" {
                        seen_from = true;
                    } else if !seen_from && node.instruction != "ARG" {
                        return Err(ParseError(format!(
                            "La instruccion '{}' aparece antes de FROM (linea {})",
                            node.instruction, tok.line
                        )));
                    }
                    ast.push(node);
                    continue;
                }
                TokenType::Newline => {
                    self.advance();
                    continue;
                }
                _ => {
                    return Err(ParseError(format!(
                        "Token inesperado {:?} en linea {}",
                        tok.ttype, tok.line
                    )));
                }
            }
        }

        Ok(ast)
    }

    fn expect_optional_newline(&mut self) {
        if self.peek().ttype == TokenType::Newline {
            self.advance();
        }
    }

    fn parse_instruction(&mut self) -> Result<Node, ParseError> {
        let instr_tok = self.advance();
        let mut node = Node {
            instruction: instr_tok.value,
            args: Vec::new(),
            form: None,
        };

        if self.peek().ttype == TokenType::LBracket {
            self.parse_exec_form(&mut node)?;
        } else {
            self.parse_shell_or_kv_form(&mut node);
        }

        self.expect_optional_newline();
        Ok(node)
    }

    fn parse_exec_form(&mut self, node: &mut Node) -> Result<(), ParseError> {
        self.advance(); // consume '['
        while self.peek().ttype != TokenType::RBracket {
            let tok = self.peek().clone();
            match tok.ttype {
                TokenType::String => {
                    node.args.push(tok.value);
                    self.advance();
                }
                TokenType::Comma | TokenType::Newline => {
                    self.advance();
                }
                _ => {
                    return Err(ParseError(format!(
                        "Token inesperado {:?} dentro de forma exec, linea {}",
                        tok.ttype, tok.line
                    )));
                }
            }
        }
        self.advance(); // consume ']'
        node.form = Some("exec".to_string());
        Ok(())
    }

    fn parse_shell_or_kv_form(&mut self, node: &mut Node) {
        let mut is_kv = false;

        loop {
            match self.peek().ttype {
                TokenType::Identifier | TokenType::Equals | TokenType::String => {
                    let tok = self.advance();
                    if tok.ttype == TokenType::Equals {
                        is_kv = true;
                        continue;
                    }
                    node.args.push(tok.value);
                }
                _ => break,
            }
        }

        node.form = Some(if is_kv { "key_value".to_string() } else { "shell".to_string() });
    }
}

pub fn parse_dockerfile(text: &str) -> Result<Vec<Node>, ParseError> {
    let tokens = tokenize(text);
    let mut parser = Parser::new(tokens);
    parser.parse()
}
